In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.7 MB 1.9 MB/s 
     |████████████████████████████████| 49 kB 5.7 MB/s 


In [2]:
from pyomo.environ import * 

In [3]:
import numpy as np

In [4]:
import pandas as pd

Answer1:
CityScape Constructions Inc. has to build n ∈ N different types of factories at n type of different locations.Assume that variable $ x_{ij}  $ denotes ith factory and jth loaction.
If $x_{ij}$ = 1 it means that $i^{th}$ factory is located at $j^{th}$ location.
and if $ x_{ij}$ =0 it means that  $i^{th}$ factory is not located at $j^{th}$ location.
So our variable takes values as binary.

Since it is given that each factory must be located at one location so:
$ \sum_{j=1}^nx_{ij}=1  $ for all i=1,2,$\cdots$n

We are also given that each location must be alloted a facory to be located so:
$ \sum_{i=1}^nx_{ij}=1  $ for all j=1,2,$\cdots ,n$ \\
$ x_{ij}\geq 0 $ ∀ i,j∈{1,2,3,...,n}

Let we have a matrix A$_{ij}$ = [$a_{ij}$] $ ∀ i,j \in \{1,2,3,\cdots,n\} $ and $a_{ij}$ denotes the cost to lacate $i^{th}$ factory at $j^{th}$ location.
Objective function: $ A_{ij} $[x$_{ij}$] 

We are to minimize this cost function.



In [5]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 155569 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [6]:
opt_cbc = SolverFactory('cbc')

First we define a function for n,data,d for assignment of factories to locations. Here n is the number of factories and corresponding number of locations.And data is representing the array obtained from .txt file which contains the cost for allocation of factories to various lacations and d represents the domain of variables. 

In [14]:
def assignment(n,data,d):
  row_indices=range(n)
  col_indices=range(n)
  model=ConcreteModel()
  model.x=Var(row_indices,col_indices,domain=d)
  model.objective = Objective(expr=sum(int(data[i+1,j+1])*model.x[i,j] for i in row_indices for j in col_indices),sense=minimize)
  model.constraints = ConstraintList()
  for j in range(n):
    model.constraints.add(expr=sum(model.x[i,j] for i in row_indices) == 1)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[i,j] for j in col_indices) == 1)  
  result = opt_cbc.solve(model)
  print('Solver status:', result.solver.status)
  print('Solver termination condition:',result.solver.termination_condition) 
  print('\nObjective = ', model.objective())

  print('\nDecision Variables')
  for i in row_indices:
    for j in col_indices:
      print(f'whether factory {i+1} is assigned to location {j+1}:', model.x[i,j].value)

  print('\nConstraints')
  model.constraints.display()
  for i in range(n):
    for j in range(n):
      if model.x[i,j].value == 1:
        print(f'factory {i+1} is assigned to location {j+1}')
        
    
      
    

Answer8:
So now for given data n=12 and data is in lab6_ex2_1.txt file which contains the cost of setting up various factories to various locations.First we take domain as Binary.

In [15]:
data_1=np.loadtxt("lab6_ex2_1.txt",dtype = 'str')

In [16]:
data_1

array([['Factory', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
        '11', '12'],
       ['1', '21', '25', '18', '19', '23', '24', '16', '17', '21', '23',
        '19', '21'],
       ['2', '19', '22', '19', '19', '23', '21', '15', '20', '16', '17',
        '17', '24'],
       ['3', '21', '19', '20', '18', '22', '24', '14', '16', '18', '16',
        '20', '24'],
       ['4', '19', '22', '17', '17', '22', '22', '18', '18', '20', '21',
        '23', '23'],
       ['5', '18', '22', '15', '15', '23', '23', '21', '25', '24', '19',
        '21', '23'],
       ['6', '23', '19', '19', '14', '24', '21', '23', '21', '20', '20',
        '22', '19'],
       ['7', '21', '16', '27', '17', '26', '27', '16', '19', '24', '20',
        '20', '23'],
       ['8', '22', '17', '21', '15', '23', '24', '24', '20', '22', '19',
        '23', '21'],
       ['9', '23', '24', '18', '19', '22', '23', '21', '18', '23', '15',
        '25', '21'],
       ['10', '18', '21', '21', '18', '26', '24', '18', '17', '

In [17]:
assignment(12,data_1,Binary)


Solver status: ok
Solver termination condition: optimal

Objective =  204.0

Decision Variables
whether factory 1 is assigned to location 1: 0.0
whether factory 1 is assigned to location 2: 0.0
whether factory 1 is assigned to location 3: 0.0
whether factory 1 is assigned to location 4: 0.0
whether factory 1 is assigned to location 5: 0.0
whether factory 1 is assigned to location 6: 0.0
whether factory 1 is assigned to location 7: 0.0
whether factory 1 is assigned to location 8: 0.0
whether factory 1 is assigned to location 9: 0.0
whether factory 1 is assigned to location 10: 0.0
whether factory 1 is assigned to location 11: 1.0
whether factory 1 is assigned to location 12: 0.0
whether factory 2 is assigned to location 1: 0.0
whether factory 2 is assigned to location 2: 0.0
whether factory 2 is assigned to location 3: 0.0
whether factory 2 is assigned to location 4: 0.0
whether factory 2 is assigned to location 5: 0.0
whether factory 2 is assigned to location 6: 0.0
whether factory 2 i

Answer8:
Minimum objective cost: 204 \\
factory 1 is assigned to location 11 \\
factory 2 is assigned to location 9 \\
factory 3 is assigned to location 8 \\
factory 4 is assigned to location 6 \\
factory 5 is assigned to location 3 \\
factory 6 is assigned to location 12 \\
factory 7 is assigned to location 7 \\
factory 8 is assigned to location 4 \\
factory 9 is assigned to location 10 \\
factory 10 is assigned to location 1 \\
factory 11 is assigned to location 2 \\
factory 12 is assigned to location 5 \\

Answer9:
Now we change the domain as NonNegativeReals.now domain is continuous:


In [18]:
assignment(12,data_1,NonNegativeReals)

Solver status: ok
Solver termination condition: optimal

Objective =  204.0

Decision Variables
whether factory 1 is assigned to location 1: 0.0
whether factory 1 is assigned to location 2: 0.0
whether factory 1 is assigned to location 3: 0.0
whether factory 1 is assigned to location 4: 0.0
whether factory 1 is assigned to location 5: 0.0
whether factory 1 is assigned to location 6: 0.0
whether factory 1 is assigned to location 7: 0.0
whether factory 1 is assigned to location 8: 0.0
whether factory 1 is assigned to location 9: 0.0
whether factory 1 is assigned to location 10: 0.0
whether factory 1 is assigned to location 11: 1.0
whether factory 1 is assigned to location 12: 0.0
whether factory 2 is assigned to location 1: 0.0
whether factory 2 is assigned to location 2: 0.0
whether factory 2 is assigned to location 3: 0.0
whether factory 2 is assigned to location 4: 0.0
whether factory 2 is assigned to location 5: 0.0
whether factory 2 is assigned to location 6: 0.0
whether factory 2 i

Answer9:

factory 1 is assigned to location 11 \\
factory 2 is assigned to location 9 \\
factory 3 is assigned to location 8 \\
factory 4 is assigned to location 6 \\
factory 5 is assigned to location 3 \\
factory 6 is assigned to location 12 \\
factory 7 is assigned to location 7 \\
factory 8 is assigned to location 4 \\
factory 9 is assigned to location  10 \\
factory 10 is assigned to location 1 \\
factory 11 is assigned to location 2 \\
factory 12 is assigned to location 5 \\

Answer10:
Yes, we get same optimal solution in both the condtions when domain is integer(Binary) and continuous real(non negative real).

Yes,the values of integers are still integer valued.It is because since we know that optimal solution is only obtained at extreme points.Since here the feasible region of Mixedinteger solution and pureintegersolution is same.Here we get the same solution.It is not necessary to get same solution in both solutions but here we get.

Answer11:
Since we are given that 
facility 3 cannot be assigned to location 4, facility
9 cannot be assigned to location 6 and facility 4 cannot be assigned to location 10. Again we increase the cost of these there assignments and make the cost much larger than others.Since we are minimize the total assignment cost so much larger values doesn't allow to be assigned there facory to corresponding location. So we make changes in .txt file and make a new file with cost of these assignments 10000.

In [19]:
data_2 = np.loadtxt('lab6_ex2_2.txt',dtype='str')

In [20]:
assignment(12,data_2,Binary)

Solver status: ok
Solver termination condition: optimal

Objective =  204.0

Decision Variables
whether factory 1 is assigned to location 1: 0.0
whether factory 1 is assigned to location 2: 0.0
whether factory 1 is assigned to location 3: 0.0
whether factory 1 is assigned to location 4: 0.0
whether factory 1 is assigned to location 5: 0.0
whether factory 1 is assigned to location 6: 0.0
whether factory 1 is assigned to location 7: 0.0
whether factory 1 is assigned to location 8: 0.0
whether factory 1 is assigned to location 9: 0.0
whether factory 1 is assigned to location 10: 0.0
whether factory 1 is assigned to location 11: 1.0
whether factory 1 is assigned to location 12: 0.0
whether factory 2 is assigned to location 1: 0.0
whether factory 2 is assigned to location 2: 0.0
whether factory 2 is assigned to location 3: 0.0
whether factory 2 is assigned to location 4: 0.0
whether factory 2 is assigned to location 5: 0.0
whether factory 2 is assigned to location 6: 0.0
whether factory 2 i

Answer11: 
Minimum Cost : 204
factory 1 is assigned to location 11 \\
factory 2 is assigned to location 9 \\
factory 3 is assigned to location 8 \\
factory 4 is assigned to location 6 \\
factory 5 is assigned to location 3 \\
factory 6 is assigned to location 12 \\
factory 7 is assigned to location 7 \\
factory 8 is assigned to location 4 \\
factory 9 is assigned to location 10 \\
factory 10 is assigned to location 1 \\
factory 11 is assigned to location 2 \\
factory 12 is assigned to location 5 \\
* Part2:
Here we get the same solution as we get in first condition when we don't restrict that  facility 3 cannot be assigned to location 4, facility
9 cannot be assigned to location 6 and facility 4 cannot be assigned to location 10.It is because when there are no restricton we got an optimal solution and in that solution we didn't assigned  facility 3 to location 4 ,facility 9 to location 6 and facility 4 to location 10.So if we restrict these conditions then it does not make any difference in optimal solution.SO we get same solution.

